### PR TITLE
Fix a report preview in the comments of one expense report.

### PR DESCRIPTION
**Pull request by wildan-m**
@Pujan92 @deetergp 

### Details

Fix a report preview in the comments of one expense report.

### Fixed Issues

$ https://github.com/Expensify/App/issues/43034
PROPOSAL: https://github.com/Expensify/App/issues/43034#issuecomment-2183633570 (Alternative)

### Tests

1. Have a HT account as admin
2. Create a policy with some violation (category required) and leave the submission frequency as instantly
3. Invite a member A
4. As a member A create a report with only one expense
5. Log in as HT admin
6. Open the report
7. Verify that expense report preview should not appear in the expense details page
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Have a HT account as admin
2. Create a policy with some violation (category required) and leave the submission frequency as instantly
3. Invite a member A
4. As a member A create a report with only one expense
5. Log in as HT admin
4. Make network offline
7. Open the expense report created from 4th step.
8. The page will be blank without inner expense report preview
9. Make network online
10. Verify that expense details showing and expense report preview should not appear in the expense details page

### QA Steps

Same as Test

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/11847279/4d0df47e-1bd6-4521-969a-d1c897e9592b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/11847279/0be504f6-c11e-40af-998c-6095fa4c0bf3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/11847279/5fe335b9-bb39-4137-b762-c5975189f4c9

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/11847279/4c6b7d55-92a8-489e-94e4-09a1280fd0cb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/11847279/5fe07cf9-b851-487c-8a66-904af494bcb4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/11847279/5f1eb3fe-87a2-4083-9b49-d91a365e3931

</details>

### CODE REVIEW

**Comment by Pujan92** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by Pujan92** [comment]
@wildan-m How about passing an extra param `transactionThreadReportID` to the function `getCombinedReportActions` which helps to avoid the calculation of IOU again?  With that, we only need to consider `transactionThreadReportID` to decide which report actions need to return. 

`transactionThreadReportID` is available only for the one transaction thread. 
https://github.com/Expensify/App/blob/a272c4b7b9aed9943c85868b512aae7a2766fe43/src/libs/ReportActionsUtils.ts#L953-L956

**Comment by wildan-m** [comment]
@Pujan92 I think that's a good idea. Something like this, right? I'll try to test it.

```tsx
function getCombinedReportActions(reportActions: ReportAction[], transactionThreadReportActions: ReportAction[], reportID?: string, transactionThreadReportID?: string): ReportAction[] {
    // There is a chance that transactionThreadReportActions is not loaded yet.
    // We only check for emptiness when there are multiple IOUs,
    // this will prevent the sub-report preview from being displayed when there is only one IOU.
    if (isEmptyObject(transactionThreadReportActions) && _.isEmpty(transactionThreadReportID)) {
        return reportActions;
    }

```

**Comment by wildan-m** [comment]
@Pujan92 the PR has been updated

File: src/libs/ReportActionsUtils.ts [hunk 1]
```diff
@@ -374,14 +374,21 @@ function shouldIgnoreGap(currentReportAction: ReportAction | undefined, nextRepo
  * Returns a sorted and filtered list of report actions from a report and it's associated child
  * transaction thread report in order to correctly display reportActions from both reports in the one-transaction report view.
  */
-function getCombinedReportActions(reportActions: ReportAction[], transactionThreadReportActions: ReportAction[], reportID?: string): ReportAction[] {
-    if (isEmptyObject(transactionThreadReportActions)) {
+function getCombinedReportActions(
+    reportActions: ReportAction[],
+    transactionThreadReportActions: ReportAction[],
+    reportID?: string,
+    transactionThreadReportID?: string | null,
```
Comment (Pujan92) [inline src/libs/ReportActionsUtils.ts:381]:
```suggestion
    reportActions: ReportAction[],
    transactionThreadReportID: string | null,
    transactionThreadReportActions: ReportAction[],
    reportID?: string,
```

I am more inclined towards making `transactionThreadReportID` mandatory like the way we have transactionThreadReportActions

File: src/libs/ReportActionsUtils.ts [hunk 2]
```diff
@@ -374,14 +374,21 @@ function shouldIgnoreGap(currentReportAction: ReportAction | undefined, nextRepo
...
+): ReportAction[] {
+    // There is a chance that transactionThreadReportActions is not loaded yet.
+    // We only check for emptiness when there are multiple IOUs,
+    // this will prevent the sub-report preview from being displayed when there is only one IOU.
```
Comment (Pujan92) [inline src/libs/ReportActionsUtils.ts:385]: Do you think we don't need this comment now?

File: src/pages/home/report/ReportActionsView.tsx [hunk 3]
```diff
@@ -157,8 +158,8 @@ function ReportActionsView({
     // Get a sorted array of reportActions for both the current report and the transaction thread report associated with this report (if there is one)
     // so that we display transaction-level and report-level report actions in order in the one-transaction view
     const combinedReportActions = useMemo(
-        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportActions),
-        [allReportActions, transactionThreadReportActions],
+        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportActions, reportID, transactionThreadReportID),
```
Comment (Pujan92) [inline src/pages/home/report/ReportActionsView.tsx:161]:
```suggestion
        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportID, transactionThreadReportActions),
```

We also need to do the same above change in `OptionListUtils.ts`

**Review by Pujan92 — CHANGES_REQUESTED** [review]
Thanks @wildan-m , suggested some minor changes.

File: src/libs/ReportActionsUtils.ts (hunk 2 above)
Comment (wildan-m) [reply src/libs/ReportActionsUtils.ts:385]: @Pujan92 I believe the comment is helpful as it provides a rationale for the necessity of checking for an empty `transactionThreadReportID`. Perhaps we could consider rephrasing it for clarity, any suggestion?

Comment (Pujan92) [reply src/libs/ReportActionsUtils.ts:385]: > // There is a chance that transactionThreadReportActions is not loaded yet.

I thought now that we have the `transactionThreadReportID` we don't need to rely on `transactionThreadReportActions` for that condition. If we go with that then we can omit this comment bcoz the function description is enough to convey. WDYT?

File: src/pages/home/report/ReportActionsView.tsx [hunk 4]
```diff
@@ -157,8 +158,8 @@ function ReportActionsView({
     // Get a sorted array of reportActions for both the current report and the transaction thread report associated with this report (if there is one)
     // so that we display transaction-level and report-level report actions in order in the one-transaction view
     const combinedReportActions = useMemo(
-        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportActions),
-        [allReportActions, transactionThreadReportActions],
+        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportID ?? null, transactionThreadReportActions, reportID),
```
Comment (Pujan92) [inline src/pages/home/report/ReportActionsView.tsx:161]:
```suggestion
        () => ReportActionsUtils.getCombinedReportActions(allReportActions, transactionThreadReportID ?? null, transactionThreadReportActions),
```
Let's avoid passing optional parameter `reportID` like the way how it was earlier

File: src/libs/ReportActionsUtils.ts [hunk 5]
```diff
@@ -374,14 +374,21 @@ function shouldIgnoreGap(currentReportAction: ReportAction | undefined, nextRepo
...
+function getCombinedReportActions(
+    reportActions: ReportAction[],
+    transactionThreadReportID: string | null,
+    transactionThreadReportActions: ReportAction[],
+    reportID?: string,
+): ReportAction[] {
+    // There is a chance that transactionThreadReportActions is not loaded yet.
+    // We only check for emptiness when there are multiple IOUs,
+    // this will prevent the sub-report preview from being displayed when there is only one IOU.
+    if (isEmptyObject(transactionThreadReportActions) && _.isEmpty(transactionThreadReportID)) {
```
Comment (Pujan92) [inline src/libs/ReportActionsUtils.ts:386]:
```suggestion
    if (_.isEmpty(transactionThreadReportID)) {
```

File: src/pages/home/report/ReportActionsView.tsx (hunk 4 above)
Comment (wildan-m) [reply src/pages/home/report/ReportActionsView.tsx:161]: @Pujan92 on the submitter side, not supplying report ID will create duplicate parent report previews when we create multiple IOUs.

Comment (Pujan92) [reply src/pages/home/report/ReportActionsView.tsx]: Thanks!

File: src/libs/ReportActionsUtils.ts (hunk 5 above)
Comment (wildan-m) [reply src/libs/ReportActionsUtils.ts:386]: @Pujan92, could you provide further clarification on this suggestion?

Comment (wildan-m) [reply src/libs/ReportActionsUtils.ts]: I understand, I think that makes sense. The PR has been updated.

**Review by Pujan92 — APPROVED** [review]
LGTM, Thanks! :)

**Comment by Pujan92** [comment]
@deetergp Just to inform you, We found that `transactionThreadReportID`(which is specifically created to identify one expense report) is enough to decide the report actions, so we implemented this straightforward solution instead of doing any more filtration.

**Review by deetergp — APPROVED** [review]
Code looks good and tests well 👍

**Pull request merged**